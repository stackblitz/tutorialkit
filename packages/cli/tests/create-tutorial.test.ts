import { readFileSync } from 'node:fs';
import fs from 'node:fs/promises';
import { tmpdir } from 'node:os';
import path from 'node:path';
import { execa } from 'execa';
import { afterAll, beforeAll, expect, test } from 'vitest';

// on CI on windows we want to make sure to use the same drive, so we use a custom logic
const tmpDir =
  process.platform === 'win32'
    ? path.join(path.resolve(__dirname, '../../../..'), '.tmp')
    : await fs.mkdtemp(path.join(tmpdir(), 'tk-test-'));
const baseDir = path.resolve(__dirname, '../../..');

const cli = path.join(baseDir, 'packages/cli/dist/index.js');

beforeAll(async () => {
  await fs.rm(tmpDir, { force: true, recursive: true });
  await fs.mkdir(tmpDir);
});

afterAll(async () => {
  if (process.platform !== 'win32' || !process.env.CI) {
    await fs.rm(tmpDir, { force: true, recursive: true });
  }
});

test('cannot create project without installing but with starting', async (context) => {
  const name = context.task.id;

  await expect(
    execa('node', [cli, 'create', name, '--no-install', '--start'], {
      cwd: tmpDir,
    }),
  ).rejects.toThrow('Cannot start project without installing dependencies.');
});

test('create a project', async (context) => {
  const name = context.task.id;
  const dest = path.join(tmpDir, name);

  await execa('node', [cli, 'create', name, '--no-install', '--no-git', '--defaults'], {
    cwd: tmpDir,
  });

  const projectFiles = await fs.readdir(dest, { recursive: true });

  expect(projectFiles.map(normaliseSlash).sort()).toMatchSnapshot();
});

test('create and build a project', async (context) => {
  const name = context.task.id;
  const dest = path.join(tmpDir, name);

  await execa('node', [cli, 'create', name, '--no-git', '--no-install', '--no-start', '--defaults'], {
    cwd: tmpDir,
  });

  await runPnpmInstall(dest, baseDir);

  await execa('pnpm', ['run', 'build'], {
    cwd: dest,
  });

  // remove `_astro` before taking the snapshot
  await fs.rm(path.join(dest, 'dist/_astro'), { force: true, recursive: true });

  const distFiles = await fs.readdir(path.join(dest, 'dist'), { recursive: true });

  expect(distFiles.map(normaliseSlash).sort()).toMatchSnapshot();

  // create snapshot of lesson, solution and template file reference JSONs
  const lessonJsons = distFiles.filter((file) => file.endsWith('-files.json'));
  const solutionJsons = distFiles.filter((file) => file.endsWith('-solution.json'));
  const templateJsons = distFiles.filter((file) => file.startsWith('template-') && file.endsWith('.json'));

  const contents = [...lessonJsons, ...solutionJsons, ...templateJsons].reduce((jsons, current) => {
    const fileJson = JSON.parse(readFileSync(path.join(dest, 'dist', current), 'utf8'));
    const filenames = Object.keys(fileJson);

    return { ...jsons, [current]: filenames };
  }, {});

  expect(contents).toMatchSnapshot('built project file references');
});

test('create and eject a project', async (context) => {
  const name = context.task.id;
  const dest = path.join(tmpDir, name);

  await execa('node', [cli, 'create', name, '--no-git', '--no-install', '--no-start', '--defaults'], {
    cwd: tmpDir,
  });

  await runPnpmInstall(dest, baseDir);

  await execa('node', [cli, 'eject', name, '--force', '--defaults'], {
    cwd: tmpDir,
  });

  if (process.platform !== 'win32') {
    await fs.rm(path.join(dest, 'node_modules'), { force: true, recursive: true, maxRetries: 5 });
  }

  let projectFiles = await fs.readdir(dest, { recursive: true });

  if (process.platform === 'win32') {
    projectFiles = projectFiles.filter((filePath) => !filePath.startsWith('node_modules'));
  }

  expect(projectFiles.map(normaliseSlash).sort()).toMatchSnapshot();
  expect(await fs.readFile(path.join(dest, 'astro.config.ts'), 'utf-8')).toMatchSnapshot();
});

test('create, eject and build a project', async (context) => {
  const name = context.task.id;
  const dest = path.join(tmpDir, name);

  await execa('node', [cli, 'create', name, '--no-git', '--no-install', '--no-start', '--defaults'], {
    cwd: tmpDir,
  });

  await runPnpmInstall(dest, baseDir);

  await execa('node', [cli, 'eject', name, '--force', '--defaults'], {
    cwd: tmpDir,
  });

  await execa('pnpm', ['install', '--no-frozen-lockfile'], {
    cwd: dest,
  });

  await execa('pnpm', ['run', 'build'], {
    cwd: dest,
  });

  // remove `_astro` before taking the snapshot
  await fs.rm(path.join(dest, 'dist/_astro'), { force: true, recursive: true });

  const distFiles = await fs.readdir(path.join(dest, 'dist'), { recursive: true });

  expect(distFiles.map(normaliseSlash).sort()).toMatchSnapshot();
});

test('cannot eject on an empty folder', async (context) => {
  const name = context.task.id;
  const dest = path.join(tmpDir, name);

  await fs.mkdir(dest);

  await expect(
    execa('node', [cli, 'eject', name, '--force', '--defaults'], {
      cwd: tmpDir,
    }),
  ).rejects.toThrow('package.json does not exists!');
});

test('cannot eject on a node project that is not an Astro project', async (context) => {
  const name = context.task.id;
  const dest = path.join(tmpDir, name);

  await fs.mkdir(dest);

  await fs.writeFile(path.join(dest, 'package.json'), JSON.stringify({ name: 'not-tutorialkit' }));

  await expect(
    execa('node', [cli, 'eject', name, '--force', '--defaults'], {
      cwd: tmpDir,
    }),
  ).rejects.toThrow('astro.config.ts does not exists!');
});

test('cannot eject on an astro project that is not using TutorialKit', async (context) => {
  const name = context.task.id;
  const dest = path.join(tmpDir, name);

  await fs.mkdir(dest);
  await fs.mkdir(path.join(dest, 'src'));

  await fs.writeFile(
    path.join(dest, 'package.json'),
    JSON.stringify({ name: 'astro', dependencies: { astro: '4.11.0' } }),
  );

  await fs.writeFile(
    path.join(dest, 'astro.config.ts'),
    `
    import { defineConfig } from 'astro';

    export default defineConfig({});
  `,
  );

  await expect(
    execa('node', [cli, 'eject', name, '--force', '--defaults'], {
      cwd: tmpDir,
    }),
  ).rejects.toThrow(`@tutorialkit${path.sep}astro does not exists!`);
});

test('cannot eject on an astro project that is not using TutorialKit 2', async (context) => {
  const name = context.task.id;
  const dest = path.join(tmpDir, name);

  await fs.mkdir(dest);
  await fs.mkdir(path.join(dest, 'src'));

  await fs.writeFile(
    path.join(dest, 'package.json'),
    JSON.stringify({ name: 'astro', dependencies: { astro: '4.11.0', '@tutorialkit/astro': '*' } }),
  );

  await fs.writeFile(
    path.join(dest, 'astro.config.ts'),
    `
    import { defineConfig } from 'astro';

    export default defineConfig({});
  `,
  );

  await execa('pnpm', ['install'], {
    cwd: dest,
  });

  await expect(
    execa('node', [cli, 'eject', name, '--force', '--defaults'], {
      cwd: tmpDir,
    }),
  ).rejects.toThrow(`Could not find import to '@tutorialkit/astro'`);
});

function normaliseSlash(filePath: string) {
  return filePath.replace(/\\/g, '/');
}

async function runPnpmInstall(dest: string, baseDir: string) {
  const packageJsonPath = path.join(dest, 'package.json');
  const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));

  packageJson.pnpm = {
    overrides: {
      '@astrojs/language-server': '2.14.1',
      '@tutorialkit/astro': `file:${baseDir}/packages/astro`,
      '@tutorialkit/react': `file:${baseDir}/packages/react`,
      '@tutorialkit/runtime': `file:${baseDir}/packages/runtime`,
      '@tutorialkit/theme': `file:${baseDir}/packages/theme`,
      '@tutorialkit/types': `file:${baseDir}/packages/types`,
    },
  };

  await fs.writeFile(packageJsonPath, JSON.stringify(packageJson, undefined, 2), 'utf8');

  await execa('pnpm', ['install'], {
    cwd: dest,
  });
}
