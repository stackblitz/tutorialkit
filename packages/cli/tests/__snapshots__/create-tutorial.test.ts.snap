// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create a project 1`] = `
[
  ".gitignore",
  ".vscode",
  ".vscode/extensions.json",
  ".vscode/launch.json",
  "README.md",
  "astro.config.ts",
  "icons",
  "icons/languages",
  "icons/languages/css.svg",
  "icons/languages/html.svg",
  "icons/languages/js.svg",
  "icons/languages/json.svg",
  "icons/languages/markdown.svg",
  "icons/languages/sass.svg",
  "icons/languages/ts.svg",
  "package.json",
  "public",
  "public/favicon.svg",
  "public/logo-dark.svg",
  "public/logo.svg",
  "src",
  "src/content",
  "src/content/config.ts",
  "src/content/tutorial",
  "src/content/tutorial/1-basics",
  "src/content/tutorial/1-basics/1-introduction",
  "src/content/tutorial/1-basics/1-introduction/1-welcome",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files/src",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files/src/index.js",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files/src/test",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files/src/test/bar.js",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_solution",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_solution/src",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_solution/src/index.js",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/content.md",
  "src/content/tutorial/1-basics/1-introduction/2-foo",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/bar",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/bar/styles.css",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/src",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/src/index.html",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/src/unicorn.js",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/src/windows_xp.png",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_solution",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_solution/src",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_solution/src/index.html",
  "src/content/tutorial/1-basics/1-introduction/2-foo/content.mdx",
  "src/content/tutorial/1-basics/1-introduction/3-bar",
  "src/content/tutorial/1-basics/1-introduction/3-bar/_files",
  "src/content/tutorial/1-basics/1-introduction/3-bar/_files/src",
  "src/content/tutorial/1-basics/1-introduction/3-bar/_files/src/index.html",
  "src/content/tutorial/1-basics/1-introduction/3-bar/content.md",
  "src/content/tutorial/1-basics/1-introduction/meta.md",
  "src/content/tutorial/1-basics/2-foo",
  "src/content/tutorial/1-basics/2-foo/1-welcome",
  "src/content/tutorial/1-basics/2-foo/1-welcome/content.md",
  "src/content/tutorial/1-basics/2-foo/meta.md",
  "src/content/tutorial/1-basics/meta.md",
  "src/content/tutorial/2-advanced",
  "src/content/tutorial/2-advanced/1-unicorn",
  "src/content/tutorial/2-advanced/1-unicorn/1-welcome",
  "src/content/tutorial/2-advanced/1-unicorn/1-welcome/content.md",
  "src/content/tutorial/2-advanced/1-unicorn/meta.md",
  "src/content/tutorial/2-advanced/meta.md",
  "src/content/tutorial/meta.md",
  "src/env.d.ts",
  "src/templates",
  "src/templates/default",
  "src/templates/default/package-lock.json",
  "src/templates/default/package.json",
  "src/templates/default/src",
  "src/templates/default/src/index.js",
  "src/templates/vite-app",
  "src/templates/vite-app-2",
  "src/templates/vite-app-2/.tk-config.json",
  "src/templates/vite-app-2/foo.txt",
  "src/templates/vite-app/.gitignore",
  "src/templates/vite-app/counter.js",
  "src/templates/vite-app/index.html",
  "src/templates/vite-app/javascript.svg",
  "src/templates/vite-app/main.js",
  "src/templates/vite-app/package-lock.json",
  "src/templates/vite-app/package.json",
  "src/templates/vite-app/public",
  "src/templates/vite-app/public/vite.svg",
  "src/templates/vite-app/style.css",
  "tsconfig.json",
  "uno.config.ts",
]
`;

exports[`create and build a project 1`] = `
[
  "1-basics",
  "1-basics-1-introduction-1-welcome-files.json",
  "1-basics-1-introduction-1-welcome-solution.json",
  "1-basics-1-introduction-2-foo-files.json",
  "1-basics-1-introduction-2-foo-solution.json",
  "1-basics-1-introduction-3-bar-files.json",
  "1-basics/2-foo",
  "1-basics/2-foo/1-welcome",
  "1-basics/2-foo/1-welcome/index.html",
  "1-basics/custom-chapter-slug",
  "1-basics/custom-chapter-slug/1-welcome",
  "1-basics/custom-chapter-slug/1-welcome/index.html",
  "1-basics/custom-chapter-slug/bar",
  "1-basics/custom-chapter-slug/bar/index.html",
  "1-basics/custom-chapter-slug/foo",
  "1-basics/custom-chapter-slug/foo/index.html",
  "2-advanced",
  "2-advanced/1-unicorn",
  "2-advanced/1-unicorn/1-welcome",
  "2-advanced/1-unicorn/1-welcome/index.html",
  "favicon.svg",
  "index.html",
  "logo-dark.svg",
  "logo.svg",
  "template-default.json",
  "template-vite-app-2.json",
  "template-vite-app.json",
]
`;

exports[`create and eject a project 1`] = `
[
  ".gitignore",
  ".vscode",
  ".vscode/extensions.json",
  ".vscode/launch.json",
  "README.md",
  "astro.config.ts",
  "icons",
  "icons/languages",
  "icons/languages/css.svg",
  "icons/languages/html.svg",
  "icons/languages/js.svg",
  "icons/languages/json.svg",
  "icons/languages/markdown.svg",
  "icons/languages/sass.svg",
  "icons/languages/ts.svg",
  "package.json",
  "pnpm-lock.yaml",
  "public",
  "public/favicon.svg",
  "public/logo-dark.svg",
  "public/logo.svg",
  "src",
  "src/components",
  "src/components/LoginButton.tsx",
  "src/components/Logo.astro",
  "src/components/MainContainer.astro",
  "src/components/MobileContentToggle.astro",
  "src/components/NavCard.astro",
  "src/components/NavWrapper.tsx",
  "src/components/PageLoadingIndicator.astro",
  "src/components/ResizablePanel.astro",
  "src/components/ThemeSwitch.tsx",
  "src/components/TopBar.astro",
  "src/components/TopBarWrapper.astro",
  "src/components/TutorialContent.astro",
  "src/components/WorkspacePanelWrapper.tsx",
  "src/components/setup.ts",
  "src/components/webcontainer.ts",
  "src/content",
  "src/content/config.ts",
  "src/content/tutorial",
  "src/content/tutorial/1-basics",
  "src/content/tutorial/1-basics/1-introduction",
  "src/content/tutorial/1-basics/1-introduction/1-welcome",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files/src",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files/src/index.js",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files/src/test",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_files/src/test/bar.js",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_solution",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_solution/src",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/_solution/src/index.js",
  "src/content/tutorial/1-basics/1-introduction/1-welcome/content.md",
  "src/content/tutorial/1-basics/1-introduction/2-foo",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/bar",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/bar/styles.css",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/src",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/src/index.html",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/src/unicorn.js",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_files/src/windows_xp.png",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_solution",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_solution/src",
  "src/content/tutorial/1-basics/1-introduction/2-foo/_solution/src/index.html",
  "src/content/tutorial/1-basics/1-introduction/2-foo/content.mdx",
  "src/content/tutorial/1-basics/1-introduction/3-bar",
  "src/content/tutorial/1-basics/1-introduction/3-bar/_files",
  "src/content/tutorial/1-basics/1-introduction/3-bar/_files/src",
  "src/content/tutorial/1-basics/1-introduction/3-bar/_files/src/index.html",
  "src/content/tutorial/1-basics/1-introduction/3-bar/content.md",
  "src/content/tutorial/1-basics/1-introduction/meta.md",
  "src/content/tutorial/1-basics/2-foo",
  "src/content/tutorial/1-basics/2-foo/1-welcome",
  "src/content/tutorial/1-basics/2-foo/1-welcome/content.md",
  "src/content/tutorial/1-basics/2-foo/meta.md",
  "src/content/tutorial/1-basics/meta.md",
  "src/content/tutorial/2-advanced",
  "src/content/tutorial/2-advanced/1-unicorn",
  "src/content/tutorial/2-advanced/1-unicorn/1-welcome",
  "src/content/tutorial/2-advanced/1-unicorn/1-welcome/content.md",
  "src/content/tutorial/2-advanced/1-unicorn/meta.md",
  "src/content/tutorial/2-advanced/meta.md",
  "src/content/tutorial/meta.md",
  "src/env-default.d.ts",
  "src/env.d.ts",
  "src/layouts",
  "src/layouts/Layout.astro",
  "src/pages",
  "src/pages/[...slug].astro",
  "src/pages/index.astro",
  "src/stores",
  "src/stores/auth-store.ts",
  "src/stores/theme-store.ts",
  "src/stores/view-store.ts",
  "src/styles",
  "src/styles/base.css",
  "src/styles/markdown.css",
  "src/styles/panel.css",
  "src/styles/variables.css",
  "src/templates",
  "src/templates/default",
  "src/templates/default/package-lock.json",
  "src/templates/default/package.json",
  "src/templates/default/src",
  "src/templates/default/src/index.js",
  "src/templates/vite-app",
  "src/templates/vite-app-2",
  "src/templates/vite-app-2/.tk-config.json",
  "src/templates/vite-app-2/foo.txt",
  "src/templates/vite-app/.gitignore",
  "src/templates/vite-app/counter.js",
  "src/templates/vite-app/index.html",
  "src/templates/vite-app/javascript.svg",
  "src/templates/vite-app/main.js",
  "src/templates/vite-app/package-lock.json",
  "src/templates/vite-app/package.json",
  "src/templates/vite-app/public",
  "src/templates/vite-app/public/vite.svg",
  "src/templates/vite-app/style.css",
  "src/utils",
  "src/utils/constants.ts",
  "src/utils/content",
  "src/utils/content.ts",
  "src/utils/content/default-localization.ts",
  "src/utils/content/squash.ts",
  "src/utils/logger.ts",
  "src/utils/nav.ts",
  "src/utils/routes.ts",
  "src/utils/url.ts",
  "src/utils/workspace.ts",
  "tsconfig.json",
  "uno.config.ts",
]
`;

exports[`create and eject a project 2`] = `
"import tutorialkit from '@tutorialkit/astro';
import { defineConfig } from 'astro/config';

export default defineConfig({
  devToolbar: {
    enabled: false
  },
  integrations: [tutorialkit({
    defaultRoutes: false
  })]
});"
`;

exports[`create, eject and build a project 1`] = `
[
  "1-basics",
  "1-basics-1-introduction-1-welcome-files.json",
  "1-basics-1-introduction-1-welcome-solution.json",
  "1-basics-1-introduction-2-foo-files.json",
  "1-basics-1-introduction-2-foo-solution.json",
  "1-basics-1-introduction-3-bar-files.json",
  "1-basics/2-foo",
  "1-basics/2-foo/1-welcome",
  "1-basics/2-foo/1-welcome/index.html",
  "1-basics/custom-chapter-slug",
  "1-basics/custom-chapter-slug/1-welcome",
  "1-basics/custom-chapter-slug/1-welcome/index.html",
  "1-basics/custom-chapter-slug/bar",
  "1-basics/custom-chapter-slug/bar/index.html",
  "1-basics/custom-chapter-slug/foo",
  "1-basics/custom-chapter-slug/foo/index.html",
  "2-advanced",
  "2-advanced/1-unicorn",
  "2-advanced/1-unicorn/1-welcome",
  "2-advanced/1-unicorn/1-welcome/index.html",
  "favicon.svg",
  "index.html",
  "logo-dark.svg",
  "logo.svg",
  "template-default.json",
  "template-vite-app-2.json",
  "template-vite-app.json",
]
`;
